NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +51,231 @@ function Base.show(io::IO, s::CategoryProduct)
+fusion_rule(::AbelianStyle, ::TrivialSector, c::CategoryProduct) = c
 
+# ===============================  Ordered implementation  =================================
+CategoryProduct(t::Tuple) = _CategoryProduct(t)
+CategoryProduct(cats::AbstractCategory...) = CategoryProduct(cats)
+
+categories_isequal(o1::Tuple, o2::Tuple) = (o1 == o2)
+
+categories_isless(::Tuple, ::Tuple) = throw(ArgumentError("Not implemented"))
+categories_isless(t1::T, t2::T) where {T<:Tuple} = t1 < t2
+
+categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
+
+categories_trivial(type::Type{<:Tuple}) = trivial.(fieldtypes(type))
+
+function categories_common(t1::Tuple, t2::Tuple)
+  n = min(length(t1), length(t2))
+  return t1[begin:n], t2[begin:n]
+end
+
+function categories_diff(t1::Tuple, t2::Tuple)
+  n1 = length(t1)
+  n2 = length(t2)
+  return n1 < n2 ? t2[(n1 + 1):end] : t1[(n2 + 1):end]
+end
+
+function shared_categories_fusion_rule(shared1::T, shared2::T) where {T<:Tuple}
+  fused = map(fusion_rule, shared1, shared2)
+  return recover_style(T, fused)
+end
+
+# ===========================  Dictionary-like implementation  =============================
 function CategoryProduct(nt::NamedTuple)
   categories = sort_keys(nt)
   return _CategoryProduct(categories)
 end
 
 CategoryProduct(; kws...) = CategoryProduct((; kws...))
 
+# avoid having 2 different kinds of TrivialSector: cast empty NamedTuple to Tuple{}
+CategoryProduct(::NamedTuple{()}) = CategoryProduct(())
+
 function CategoryProduct(pairs::Pair...)
   keys = ntuple(n -> Symbol(pairs[n][1]), length(pairs))
   vals = ntuple(n -> pairs[n][2], length(pairs))
   return CategoryProduct(NamedTuple{keys}(vals))
 end
 
-function categories_fusion_rule(A::NamedTuple, B::NamedTuple)
-  qs = [A]
-  for (la, lb) in zip(pairs(intersect_keys(A, B)), pairs(intersect_keys(B, A)))
-    @assert la[1] == lb[1]
-    fused_vals = ⊗(la[2], lb[2])
-    qs = [union_keys((; la[1] => v), q) for v in fused_vals for q in qs]
-  end
-  # Include sectors of B not in A
-  qs = [union_keys(q, B) for q in qs]
-  return qs
+function categories_isequal(nt::NamedTuple, ::Tuple{})
+  return categories_isequal(nt, (;))
+end
+function categories_isequal(::Tuple{}, nt::NamedTuple)
+  return categories_isequal((;), nt)
+end
+function categories_isequal(nt1::NamedTuple, nt2::NamedTuple)
+  return ==(sym_categories_insert_unspecified(nt1, nt2)...)
 end
 
-function categories_equal(A::NamedTuple, B::NamedTuple)
-  common_categories = zip(pairs(intersect_keys(A, B)), pairs(intersect_keys(B, A)))
-  common_categories_match = all(nl -> (nl[1] == nl[2]), common_categories)
-  unique_categories_zero = all(l -> istrivial(l), symdiff_keys(A, B))
-  return common_categories_match && unique_categories_zero
+function categories_isless(nt::NamedTuple, ::Tuple{})
+  return categories_isless(nt, (;))
+end
+function categories_isless(::Tuple{}, nt::NamedTuple)
+  return categories_isless((;), nt)
+end
```

Comment:
Please move these next to the generic definition `categories_isless(::Tuple, ::NamedTuple)`.